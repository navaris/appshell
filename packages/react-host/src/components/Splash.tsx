import React from 'react';
import Loading from 'react-spinners/PropagateLoader';
import env from '../runtime.env';

const Splash = () => (
  <div
    style={{
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      flexGrow: 1,
      height: '100vh',
      backgroundColor: env.APPSHELL_THEME_COLOR,
    }}
  >
    <Loading color={env.APPSHELL_PRIMARY_COLOR} speedMultiplier={1.5} size={24} />
  </div>
);

export default Splash;
