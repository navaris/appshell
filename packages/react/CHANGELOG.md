# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.2.1](https://github.com/navaris/appshell/compare/@appshell/react@0.2.0...@appshell/react@0.2.1) (2023-08-29)


### Bug Fixes

* appshell loader override check ([aedda4a](https://github.com/navaris/appshell/commit/aedda4abf81560297dce03df09fba68e8785242f))





# [0.2.0](https://github.com/navaris/appshell/compare/@appshell/react@0.2.0-alpha.0...@appshell/react@0.2.0) (2023-08-28)

**Note:** Version bump only for package @appshell/react





# [0.2.0-alpha.0](https://github.com/navaris/appshell/compare/@appshell/react@0.1.0...@appshell/react@0.2.0-alpha.0) (2023-08-23)


### Features

* consolidate runtime artifacts ([#26](https://github.com/navaris/appshell/issues/26)) ([a29479a](https://github.com/navaris/appshell/commit/a29479a49f0c5ec1273c9f8e4c7384096f2d4ba0))





# [0.1.0](https://github.com/navaris/appshell/compare/@appshell/react@0.1.0-alpha.6...@appshell/react@0.1.0) (2023-07-11)

**Note:** Version bump only for package @appshell/react





# 0.1.0-alpha.6 (2023-07-10)


### Bug Fixes

* rename @appshell/react-federated-component package to @appshell/react ([0b74498](https://github.com/navaris/appshell/commit/0b74498bf17ba73db4d936c8a012af1d3b111a28))





# [0.1.0-alpha.5](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.1.0-alpha.4...@appshell/react-federated-component@0.1.0-alpha.5) (2023-07-10)


### Bug Fixes

* registry projection, enable env overwrite, refactor docker workflow, consolidate env names ([33a28fe](https://github.com/navaris/appshell/commit/33a28fe76b58e05c5b6b6b33d4b402e52bb29e70))





# [0.1.0-alpha.4](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.1.0-alpha.3...@appshell/react-federated-component@0.1.0-alpha.4) (2023-07-06)


### Bug Fixes

* metadata urls ([efcce1c](https://github.com/navaris/appshell/commit/efcce1ce4cb1745afa83661cae1414a73f837483))





# [0.1.0-alpha.3](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.1.0-alpha.2...@appshell/react-federated-component@0.1.0-alpha.3) (2023-07-06)


### Bug Fixes

* support metadata generation ([4df202f](https://github.com/navaris/appshell/commit/4df202f0fd3b9ca6c660975b75eb0ac9b60225c2))





# [0.1.0-alpha.2](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.1.0-alpha.1...@appshell/react-federated-component@0.1.0-alpha.2) (2023-07-06)


### Bug Fixes

* lint ([a4861ab](https://github.com/navaris/appshell/commit/a4861abfe82d1e0ab815aaa62cdf7907229d0136))


### Features

* appshell start, cleanup dev workflow ([74fd2e5](https://github.com/navaris/appshell/commit/74fd2e5a5acd2415482268175c7f3f16cd7c93ec))





# [0.1.0-alpha.1](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.0.3...@appshell/react-federated-component@0.1.0-alpha.1) (2023-07-05)


### Features

* foundation refactor wip ([339b930](https://github.com/navaris/appshell/commit/339b9306accaa9aba1712f3a0dee6c4cab7ed273))





# [0.1.0-alpha.0](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.0.3...@appshell/react-federated-component@0.1.0-alpha.0) (2023-07-05)


### Features

* registry index ([8e6ee0a](https://github.com/navaris/appshell/commit/8e6ee0a6a377584efa2ee702168025f46108b8c5))





## [0.0.3](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.0.3-alpha.2...@appshell/react-federated-component@0.0.3) (2023-06-29)

**Note:** Version bump only for package @appshell/react-federated-component





## [0.0.3-alpha.2](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.0.2...@appshell/react-federated-component@0.0.3-alpha.2) (2023-06-29)

**Note:** Version bump only for package @appshell/react-federated-component






## [0.0.3-alpha.1](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.0.3-alpha.0...@appshell/react-federated-component@0.0.3-alpha.1) (2023-06-29)


### Bug Fixes

* update npm package visibility ([1ef4119](https://github.com/navaris/appshell/commit/1ef411903dd038dfc781e8ce0700811e5460c903))





## [0.0.3-alpha.0](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.0.2-alpha.1...@appshell/react-federated-component@0.0.3-alpha.0) (2023-06-29)

**Note:** Version bump only for package @appshell/react-federated-component





## [0.0.2](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.0.2-alpha.1...@appshell/react-federated-component@0.0.2) (2023-06-29)

**Note:** Version bump only for package @appshell/react-federated-component





## [0.0.2-alpha.1](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.0.2-alpha.0...@appshell/react-federated-component@0.0.2-alpha.1) (2023-06-24)


### Bug Fixes

* add useEffect dependencies for federated component ([616039e](https://github.com/navaris/appshell/commit/616039ebb3f46f34db0b82acfb6ee5438a6b4731))





## [0.0.2-alpha.0](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.0.1...@appshell/react-federated-component@0.0.2-alpha.0) (2023-06-15)


### Bug Fixes

* hot module replacement support, example updates ([a67d230](https://github.com/navaris/appshell/commit/a67d2303b40f5911373225cc0b4ccf9b67c33d11))





## 0.0.1 (2023-06-15)

**Note:** Version bump only for package @appshell/react-federated-component





## [0.0.1-alpha.2](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.0.1-alpha.1...@appshell/react-federated-component@0.0.1-alpha.2) (2023-06-15)


### Bug Fixes

* hot module replacement support, example updates ([a67d230](https://github.com/navaris/appshell/commit/a67d2303b40f5911373225cc0b4ccf9b67c33d11))





## [0.0.1-alpha.1](https://github.com/navaris/appshell/compare/@appshell/react-federated-component@0.0.1-alpha.0...@appshell/react-federated-component@0.0.1-alpha.1) (2023-03-19)

**Note:** Version bump only for package @appshell/react-federated-component





## 0.0.1-alpha.0 (2023-03-19)

**Note:** Version bump only for package @appshell/react-federated-component
