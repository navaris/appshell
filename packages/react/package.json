{
  "name": "@appshell/react",
  "version": "0.2.0",
  "description": "React utilities for building micro-frontends with Appshell and Module Federation",
  "main": "dist/main.js",
  "types": "dist/types/react/src/index.d.ts",
  "repository": "https://github.com/navaris/appshell.git",
  "author": "Robert Hamilton <rh@navaris.com>",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "scripts": {
    "build": "webpack --mode production && yarn build:types",
    "build:types": "tsc --emitDeclarationOnly",
    "clean": "rm -rf dist",
    "start": "webpack --watch --mode development --stats-error-details",
    "stats": "webpack --json > stats.json"
  },
  "keywords": [
    "webpack",
    "module federation",
    "micro-frontends",
    "appshell"
  ],
  "license": "MIT",
  "peerDependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  }
}