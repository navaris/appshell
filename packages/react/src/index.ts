export { default as FederatedComponent } from './components/FederatedComponent';
export { default as ReactHost } from './components/ReactHost';
export { GlobalConfigProvider } from './contexts/GlobalConfigContext';
export { ManifestProvider } from './contexts/ManifestContext';
export { MetadataProvider } from './contexts/MetadataContext';
export { default as useGlobalConfig } from './hooks/useGlobalConfig';
export { default as useManifest } from './hooks/useManifest';
export { default as useMetadata } from './hooks/useMetadata';
export { default as jsonResource } from './resources/jsonResource';
