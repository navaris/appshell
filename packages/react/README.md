<div align="center">
  <a href="https://github.com/navaris/appshell">
    <picture>
      <source media="(prefers-color-scheme: dark)" srcset="https://github.com/navaris/appshell/blob/main/assets/branding/appshell-logo-white_2x.png">
      <img alt="appshell" src="https://github.com/navaris/appshell/blob/main/assets/branding/appshell-logo_2x.png">
    </picture>
  </a>
</div>

[![Appshell CI](https://github.com/navaris/appshell/actions/workflows/pipeline.yml/badge.svg)](https://github.com/navaris/appshell/actions/workflows/pipeline.yml)

# @appshell/react

React utilites for building micro-frontends with Webpack Module federation and Appshell.

Working examples can be found [here](https://github.com/navaris/appshell/tree/main/examples).

## Getting Started

To begin, you'll need to install `@appshell/react`:

```console
npm install @appshell/react --save-dev
```

or

```console
yarn add -D @appshell/react
```

or

```console
pnpm add -D @appshell/react
```

## FederatedComponent

React component that dynamically loads federated components.

```tsx
import { FederatedComponent, ManifestProvider } from '@appshell/react';

<App>
  <ManifestProvider manifest={manifest}>
    <FederatedComponent remote="PingModule/Ping">
    <FederatedComponent remote="PongModule/Pong">
  </ManifestProvider>
</App>
```

## useGlobalConfig

For access to the global appshell config.

```tsx
import { GlobalConfigProvider, useGlobalConfig } from '@appshell/react';

const MyComponent = () => {
  const config = useGlobalConfig();

  ...
}

<RegistryProvider config={config}>
  <MyComponent />
</RegistryProvider>
```

## useManifest

For access to the manifest.

```tsx
import { ManifestProvider, useManifest } from '@appshell/react';

const MyComponent = () => {
  const manifest = useManifest();

  ...
}

<ManifestProvider manifest={manifest}>
  <MyComponent />
</ManifestProvider>
```

**Where does the registry come from?**

Currently, the registry is a shared directory or mounted volume. It does support calling an http endpoint, but it has not yet been implemented.

> See [@appshell/cli](https://www.npmjs.com/package/@appshell/cli)

Each micro-frontend registers its manifest with the registry

```bash
appshell register --manifest dist/appshell.manifest.json
```
