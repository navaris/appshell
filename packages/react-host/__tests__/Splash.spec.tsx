/** @jest-environment jsdom */
import { render } from '@testing-library/react';
import React from 'react';
import Splash from '../src/components/Splash';

jest.mock('../src/runtime.env', () => ({
  APPSHELL_THEME_COLOR: '#000000',
  APPSHELL_PRIMARY_COLOR: 'orangered',
}));

describe('ErrorMessage', () => {
  it('should match snapshot', () => {
    const view = render(<Splash />);

    expect(view).toMatchSnapshot();
  });
});
