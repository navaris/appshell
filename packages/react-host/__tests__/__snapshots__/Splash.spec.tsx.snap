// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="display: flex; align-items: center; justify-content: center; flex-grow: 1; height: 100vh; background-color: rgb(0, 0, 0);"
      >
        <span
          style="display: inherit; position: relative;"
        >
          <span
            style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-0 1s infinite; animation-fill-mode: forwards;"
          />
          <span
            style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-1 1s infinite; animation-fill-mode: forwards;"
          />
          <span
            style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-2 1s infinite; animation-fill-mode: forwards;"
          />
          <span
            style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-3 1s infinite; animation-fill-mode: forwards;"
          />
          <span
            style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-4 1s infinite; animation-fill-mode: forwards;"
          />
          <span
            style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-5 1s infinite; animation-fill-mode: forwards;"
          />
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="display: flex; align-items: center; justify-content: center; flex-grow: 1; height: 100vh; background-color: rgb(0, 0, 0);"
    >
      <span
        style="display: inherit; position: relative;"
      >
        <span
          style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-0 1s infinite; animation-fill-mode: forwards;"
        />
        <span
          style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-1 1s infinite; animation-fill-mode: forwards;"
        />
        <span
          style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-2 1s infinite; animation-fill-mode: forwards;"
        />
        <span
          style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-3 1s infinite; animation-fill-mode: forwards;"
        />
        <span
          style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-4 1s infinite; animation-fill-mode: forwards;"
        />
        <span
          style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-5 1s infinite; animation-fill-mode: forwards;"
        />
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
